src/main/java/org/junitpioneer/jupiter/report/PioneerReportListener.java

```diff
@@ -0,0 +1,220 @@
+	 */
+	Testcase processTest(TestIdentifier test) {
+		String status = testStatusCache.getOrDefault(test.getUniqueId(), "UNKNOWN");
+		Testcase testcase = new Testcase(test.getDisplayName(), status);
+
+		// Add properties, if there are any
+		Properties allProps = getProperties(test.getUniqueId());
+
+		if (!allProps.getProperty().isEmpty()) {
+			testcase.setProperties(allProps);
+		}
+
+		return testcase;
+	}
+
+	/**
+	 * Retrieves all cached properties for an {@code TestIdentifier} by its unique id.
+	 *
+	 * @param testIdentifierUniqueId Unique id the properties should be loaded for.
+	 * @return Wrapper which contains all cached properties or an empty list.
+	 */
+	Properties getProperties(String testIdentifierUniqueId) {
+		Properties allProperties = new Properties();
+
+		if (propertyCache.containsKey(testIdentifierUniqueId)) {
+			allProperties.getProperty().addAll(propertyCache.get(testIdentifierUniqueId));
+		}
+
+		return allProperties;
+	}
+
+	/**
+	 * Creates the report and validates it against the XSD.
+	 *
+	 * @param report Report to be created.
+	 */
+	void writeReport(PioneerReport report) {
+
+		try {
+			// Create report file (delete first, if already exists)
+			// Target directory is "./build/reports/pioneerreport.xml"
+			String currentDir = System.getProperty("user.dir");
+			Path xmlFile = Paths.get(currentDir, "build", "reports", "pioneerreport.xml");
+			Files.deleteIfExists(xmlFile);
+			Files.createDirectories(xmlFile.getParent());
+			Files.createFile(xmlFile);
+
+			// Marshal with options
+			JAXBContext jaxbContext = JAXBContext.newInstance(PioneerReport.class);
+			Marshaller jaxbMarshaller = jaxbContext.createMarshaller();
+			jaxbMarshaller.setProperty(Marshaller.JAXB_FORMATTED_OUTPUT, Boolean.TRUE);
+			jaxbMarshaller.setProperty(Marshaller.JAXB_ENCODING, "UTF-8");
+			jaxbMarshaller.marshal(report, xmlFile.toFile());
+
+			// Validate
+			Source xmlFileSource = new StreamSource(xmlFile.toFile());
+			SchemaFactory schemaFactory = SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI);
+			File xsdFile = new File(ClassLoader.getSystemResource("xsd/pioneerreport.xsd").toURI());
+			Schema schema = schemaFactory.newSchema(xsdFile);
+			Validator validator = schema.newValidator();
+			validator.validate(xmlFileSource);
+		}
+		catch (JAXBException | IOException | SAXException | URISyntaxException e) {
+			// Throw RuntimeException to break execution regardless of caller
+			throw new RuntimeException("Error while creating the pioneer report", e); //NOSONAR
+		}
+	}
+
+	/*
+	 Implementation note:
+	 Package private for test purpose, because the listener overwrites its own file.
+	 When testing the listener itself the report is created the first time (by the tested listener)
+	 and then overwritten by the registered listener after all tests are done.
+	 A system property to configure the report name from outside was seen as too much overhead by the author.
+	 */
+	PioneerReport getPioneerReport() {
```

Comment:
This could be `public`. Is there a reason to 'hide' this?

Reply:
This method only exists for tests and if it would be possible I won't have created it at all.